extensions/amp-user-location/0.1/config-loader.js

```diff
@@ -0,0 +1,199 @@
+ */
+
+import {Deferred} from '../../../src/utils/promise';
+import {
+  UrlReplacementPolicy,
+  batchFetchJsonFor,
+} from '../../../src/batched-json';
+import {dict} from '../../../src/utils/object';
+import {isJsonScriptTag} from '../../../src/dom';
+import {isObject} from '../../../src/types';
+import {tryParseJson} from '../../../src/json';
+import {user, userAssert} from '../../../src/log';
+
+const TAG = 'amp-user-location';
+
+/**
+ * ConfigLoader unifies configuration parsing logic for local and remote
+ * config resources.
+ */
+export class ConfigLoader {
+  /**
+   * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   * @param {!Element} element
+   */
+  constructor(ampdoc, element) {
+    /** @private @const */
+    this.ampdoc_ = ampdoc;
+
+    /** @private @const */
+    this.element_ = element;
+
+    /**
+     * This is not intended to be referenced directly.
+     * Instead, call `getConfig_`
+     * @private {?Promise<!JsonObject>}
+     */
+    this.fetchPromise_ = null;
+
+    /**
+     * This is not intended to be referenced directly.
+     * Instead, call `getConfig_`
+     * @private {?Promise<!JsonObject>}
+     */
+    this.scriptConfigPromise_ = null;
+  }
+
+  /**
+   * Loads the local or remote config, if available. There are 4 cases:
+   * 1. Local config only
+   * 2. Local and remote config, with local used before remote config loads
+   * 3. Remote config only
+   * 4. No config specified
+   * @return {!Promise<!JsonObject>}
+   */
+  getConfig() {
+    // If there is a local config but not remote config,
+    // then use the local config immediately.
+    if (this.hasScriptConfig_() && !this.hasRemoteConfig_()) {
+      return this.getScriptConfig_();
+    }
+
+    if (this.hasScriptConfig_() && this.hasRemoteConfig_()) {
+      // Prefer the fetchConfig if it's ready, and if not use the script config.
+      return Promise.race([this.fetchConfig(), this.getScriptConfig_()]);
+    }
+
+    // this.hasScriptConfig_() must have returned false by this point
+    if (this.hasRemoteConfig_()) {
+      return this.fetchConfig();
+    }
+
+    // If there is no remote or local config, use an empty dict.
+    return Promise.resolve(dict());
+  }
+
+  /**
+   * @return {boolean}
+   * @private
+   */
+  hasScriptConfig_() {
+    return this.element_.children.length > 0;
+  }
+
+  /**
+   * @return {boolean}
+   * @private
+   */
+  hasRemoteConfig_() {
+    return this.element_.hasAttribute('src');
+  }
+
+  /**
+   * Parse a JSON script tag for configuration, if present.
+   * @return {!Promise<!JsonObject>}
+   * @private
+   */
+  getScriptConfig_() {
+    if (!this.hasScriptConfig_()) {
+      return Promise.resolve(dict());
+    }
+
+    if (this.scriptConfigPromise_) {
+      return this.scriptConfigPromise_;
+    }
+
+    const {children} = this.element_;
+    userAssert(
+      children.length == 1,
+      'amp-user-location may only have one configuration json <script> tag'
```

Comment:
This class seems mostly generic. Is it intended to be generic for other components to use in the future?

Seems like it could be. Passing in the tag name for the error into the constructor of the class would be a small refactor that separates out concerns a bit better.

Reply:
I did write it with the intention of it being as generic as possible, but I considered waiting for it to get factored out to make the logging messages generic. I'll do it now instead